SYCL/ESIMD/api/functional/operators/operator_decrement_and_increment.hpp

```diff
@@ -0,0 +1,299 @@
+public:
+  IncrementAndDecrementTestDescription(size_t index, DataT retrieved_val,
+                                       DataT expected_val,
+                                       const std::string &error_details,
+                                       const std::string &data_type)
+      : m_data_type(data_type), m_retrieved_val(retrieved_val),
+        m_expected_val(expected_val), m_index(index),
+        m_error_details(error_details) {}
+
+  std::string to_string() const override {
+    std::string log_msg("Failed for simd<");
+
+    log_msg += m_data_type + ", " + std::to_string(NumElems) + ">";
+    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
+    log_msg += ", expected: " + std::to_string(m_expected_val);
+    log_msg += ", at index: " + std::to_string(m_index);
+    log_msg += " for " + TestCaseT::get_description() + " operator: ";
+    log_msg += m_error_details;
+
+    return log_msg;
+  }
+
+private:
+  const std::string m_data_type;
+  const DataT m_retrieved_val;
+  const DataT m_expected_val;
+  const size_t m_index;
+  const std::string m_error_details;
+};
+
+struct base_test {
+  template <typename DataT, int NumElems, typename TestCaseT>
+  static std::vector<DataT> generate_input_data() {
+    std::vector<DataT> ref_data = generate_ref_data<DataT, NumElems>();
+
+    if constexpr (TestCaseT::is_increment()) {
+      mutate(ref_data, mutator::For_addition<DataT>(1));
+    } else if constexpr (!TestCaseT::is_increment()) {
+      mutate(ref_data, mutator::For_subtraction<DataT>(1));
+    }
+
+    return ref_data;
+  }
+};
+
+struct fp_accuracy_test {
+  template <typename DataT, int NumElems, typename TestCaseT>
+  static std::vector<DataT> generate_input_data() {
+    std::vector<DataT> ref_data;
+
+    static const DataT min = value<DataT>::lowest();
+    static const DataT denorm_min = value<DataT>::denorm_min();
+    static const DataT max = value<DataT>::max();
+    static const DataT inexact = static_cast<DataT>(0.1);
+
+    if constexpr (TestCaseT::is_increment()) {
+      ref_data.reserve((NumElems > 1) ? NumElems : 6);
+      ref_data.insert(ref_data.end(),
+                      {inexact, denorm_min, -denorm_min,
+                       value<DataT>::pos_ulp(static_cast<DataT>(-1.0)),
+                       value<DataT>::pos_ulp(min),
+                       value<DataT>::neg_ulp(max - 1)});
+      for (size_t i = ref_data.size(); i < NumElems; ++i) {
+        ref_data.push_back(inexact * i);
+      }
+    } else if constexpr (!TestCaseT::is_increment()) {
+      ref_data.reserve((NumElems > 1) ? NumElems : 6);
+      ref_data.insert(ref_data.end(),
+                      {inexact, denorm_min, -denorm_min,
+                       value<DataT>::neg_ulp(static_cast<DataT>(-1.0)),
+                       value<DataT>::neg_ulp(max),
+                       value<DataT>::pos_ulp(min + 1)});
+      for (size_t i = ref_data.size(); i < NumElems; ++i) {
+        ref_data.push_back(inexact * i);
+      }
```

Comment:
Maybe we might move the same blocks out of if-else branches?
```
if () {
} else {
}
// Fill the reference data to be no less than simd vector length
...
return ref_dat;
```

Reply:
fixed [here](https://github.com/intel/llvm-test-suite/pull/827/commits/7dddeb4523bc5d5bf8bb785d6f17a1c6e2461de4#diff-957362298136f7a2e0ab66c1cff4239caca52ea84dab6e6b2728ccc9411ebe0eL181-R191)